src/basics.ml

```diff
@@ -216,3 +216,55 @@ let term_of_rhs : rule -> term = fun r ->
     TE_Some(Bindlib.unbox (Bindlib.bind_mvar vars p))
   in
   Bindlib.msubst r.rhs (Array.mapi fn r.vars)
+
+(** {3 Subterm position helper module} *)
+
+(** Some tools to encode the position of a subterm in a term. *)
+module Subterm =
+struct
+  (** Each element of the list is a level in the tree of the term.  For
+      instance, in the term [Appl(Appl(f, x), Appl(Appl(g, a), b))], the
+      subterm [a] has position [1.0], encoded by [[0 ; 1]], [b] has [1.1]
+      encoded by [[1 ; 1]] and [x] has [0] encoded by [[0]]. *)
+  type t = int list
+
+  (** [compare a b] implements lexicographic order on positions. *)
+  let compare : t -> t -> int = fun a b -> Pervasives.compare (List.rev a)
+    (List.rev b)
+
+  (** [pp o p] output position [p] to channel [o]. *)
+  let pp : t pp = fun oc pos ->
+    match pos with
+    | [] -> Format.fprintf oc "ε"
+    | x  -> List.pp (fun oc -> Format.fprintf oc "%d") "." oc (List.rev x)
+
+  (** Initial position. *)
+  let init = []
+
+  (** [succ p] returns the successor of position [p].  For instance, if
+      [p = [1 ; 1]], [succ p = [2 ; 1]].  The succ of the initial position is
+      the first subterm of this position. *)
+  let succ = function
+    | []      -> 0 :: init
+    | x :: xs -> succ x :: xs
+
+  (** [prefix p q] sets position [p] as prefix of position [q], for instance,
+      {i prefix 1 3.4} is {i 1.3.4}; which is represented [prefix [1] [4;3]]
+      is [[4;3;1]]. *)
+  let prefix : t -> t -> t = fun p q -> q @ p
+
+  (** [sub p] returns the position of the first subterm of [p]. *)
+  let sub : t -> t = fun p -> 0 :: p
+
+  (** [tag ?s l] attaches the positions to a list of terms as if they were the
+      subterms of a same term.  If [?s] is supplied, the first element of the
+      list is at position [s]. *)
+  let tag : ?ini:t -> term list -> (term * t) list = fun ?ini xs ->
```

Comment:
Please use the same names in your comment and in the actual definition of the function.

Reply:
Rename ini into something else as it may easily be confused with init which is used a few lines below.

Reply:
Or why not renaming init into empty instead? That looks better.

Reply:
Why not. Done in 5bf5956.